File: src/main/java/com/back/nbe9112team06/global/exception/BusinessException.java
package com.back.nbe9112team06.global.exception;

import com.back.nbe9112team06.global.error.ErrorCode;
import lombok.Getter;
import org.springframework.http.HttpStatus;
import org.springframework.http.ProblemDetail;

@Getter
public class BusinessException extends RuntimeException {
    private final ErrorCode errorCode;

    public BusinessException(ErrorCode errorCode) {
        super(errorCode.getMessage());
        this.errorCode = errorCode;
    }

    public BusinessException(ErrorCode errorCode, String customMessage) {
        super(customMessage);
        this.errorCode = errorCode;
    }

    public HttpStatus getHttpStatus() {
        return errorCode.getStatus();
    }

    public String getErrorCode() {
        return errorCode.getCode();
    }

    @Override
    public String getMessage() {
        return super.getMessage() != null ? super.getMessage() : errorCode.getMessage();
    }

    /**
     * ProblemDetail 로 변환 (GlobalExceptionHandler 에서 사용)
     */
    public ProblemDetail toProblemDetail(String path) {
        return errorCode.toProblemDetail(this.getMessage(), path);
    }
}

File: src/main/java/com/back/nbe9112team06/global/exception/BusinessException.kt
package com.back.nbe9112team06.global.exception

import com.back.nbe9112team06.global.error.ErrorCode
import org.springframework.http.HttpStatus
import org.springframework.http.ProblemDetail
/***
 *  getErrorCode()가 val 자동생성된 getErrorCode()과
 *  errorCode에 있는 Code을 가져오는 getErrorCode()와 2가지가 있었어 문제 발생
 *  private로 공개 getter 차단
***/

class BusinessException(
    private val errorCode: ErrorCode,
    customMessage: String?
) : RuntimeException(customMessage) {
    constructor(errorCode: ErrorCode): this(errorCode, null)


    fun getErrorCode(): String = errorCode.code

    val httpStatus: HttpStatus
        get() = errorCode.status

    override val message: String?
        get() = super.message ?: errorCode.message

    /**
     * ProblemDetail 로 변환 (GlobalExceptionHandler 에서 사용)
     */
    fun toProblemDetail(path: String): ProblemDetail {
        val detail = message ?: errorCode.message
        return errorCode.toProblemDetail(detail, path)
    }
}